### PR TITLE
LXD Profile force

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

The following passes force as an option to the deployment info, so
that we can verify if we need to show an error or not when
validating a deploy.

The force flag overrides the error, but still validates the charm,
with the aim to improve error messages when attempting to debug
a deployment from the logs.

## QA steps

```sh
juju bootstrap lxd-test
juju deploy cs:~juju-qa/bionic/lxd-profile-fail-0 --force
```

## Documentation changes

TBA

## Bug reference

N/A

### CODE REVIEW

File: cmd/juju/application/validate.go [hunk 1]
```diff
@@ -30,13 +27,19 @@ func (r *ValidateLXDProfileCharm) SetPlanURL(planURL string) {
...
-			if err := lxdprofile.ValidateCharmInfoLXDProfile(charmInfo); err != nil {
-				return errors.Trace(err)
+	// if the charm info is not empty, we should use that to validate the
+	// lxd profile.
+	if charmInfo := deployInfo.CharmInfo; charmInfo != nil {
+		if err := lxdprofile.ValidateCharmInfoLXDProfile(charmInfo); err != nil {
+			// The force flag was provided, but we should let the user know that
+			// this could deliver some unexpected results.
+			if deployInfo.Force {
+				// TODO (stickupkid): should we consider raising this to the user
```
Comment (SimonRichardson) [inline cmd/juju/application/validate.go]: @hmlanigan do you think we should notify a user about using `--force` when a user tries to deploy a potentially invalid lxd profile?

Comment (hmlanigan) [reply cmd/juju/application/validate.go]: yes

**Comment by hmlanigan** [comment]
@SimonRichardson , @mitechie 
I'm wondering if we should follow the lead of juju upgrade-charm and have 2 forces: --force (--force-series) and --force-lxd-profile?  Rather than combining.

**Comment by SimonRichardson** [comment]
New flag description that should be changed to:
 
> Allow a charm to be deployed which bypasses checks such as supported series or LXD profile whitelist

File: cmd/juju/application/deploy.go [hunk 2]
```diff
@@ -662,9 +663,6 @@ func (c *DeployCommand) SetFlags(f *gnuflag.FlagSet) {
 }
 
 func (c *DeployCommand) Init(args []string) error {
-	if c.Force && c.Series == "" && c.PlacementSpec == "" {
```
Comment (hmlanigan) [inline cmd/juju/application/deploy.go]: Do we know why PlacementSpec is in this list?

Comment (SimonRichardson) [reply cmd/juju/application/deploy.go]: Not a clue - it's not tested :disappointed:

**Comment by SimonRichardson** [comment]
https://trello.com/c/oLR5ZYKg/498-allow-bypassing-the-validation-step-if-force-is-used

**Comment by SimonRichardson** [comment]
Example of `--force` status output

```sh
$ juju status                                          
Model    Controller  Cloud/Region         Version      SLA          Timestamp
default  xxx         localhost/localhost  2.5-beta1.1  unsupported  15:25:29+01:00

App               Version  Status   Scale  Charm             Store  Rev  OS      Charm version  Notes
lxd-profile       18.04    active       1  lxd-profile       local    0  ubuntu                 
lxd-profile-fail           waiting    0/1  lxd-profile-fail  local    0  ubuntu                 

Unit                Workload  Agent       Machine  Public address  Ports  Message
lxd-profile-fail/0  waiting   allocating  0                               waiting for machine
lxd-profile/0*      active    idle        1        10.88.71.14            load: 1.30, 2.12, 1.51

Machine  State    DNS          Inst id        Series  AZ  Message
0        down                  pending        bionic      cannot write lxd profiles: Invalid device type for device 'bdisk'
1        started  10.88.71.14  juju-7e24a8-1  bionic      Running
```

**Comment by mitechie** [comment]
This is just a failure because the ask in the profile is not legit for the machine. If it was to do something that could be allowed it'd not have any error in this output correct?

File: juju/testing/conn.go [hunk 3]
```diff
@@ -12,6 +12,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/juju/juju/core/lxdprofile"
```
Comment (hmlanigan) [inline juju/testing/conn.go]: nit: placement

File: cmd/juju/application/upgradecharm.go [hunk 4]
```diff
@@ -211,6 +212,7 @@ func (c *upgradeCharmCommand) Info() *cmd.Info {
 
 func (c *upgradeCharmCommand) SetFlags(f *gnuflag.FlagSet) {
 	c.ModelCommandBase.SetFlags(f)
+	f.BoolVar(&c.Force, "force", false, "Allow a charm to be deployed which bypasses for LXD profile allow list")
```
[the remark this answers is not in the document]
Comment (mitechie) [reply cmd/juju/application/upgradecharm.go]: I'd like to just move towards --force as the single hammer vs a hammer for each thing. Can we add a follow up card that --force will do each of the individual force options but leave the old flags as backward compatibility?

**Comment by hmlanigan** [comment]
QA:
$  juju deploy  ./testcharms/charm-repo/quantal/lxd-profile lxd-profile-local; 
  Deploying charm "local:bionic/lxd-profile-0"
$ juju deploy  ./testcharms/charm-repo/quantal/lxd-profile-fail lxd-profilefail-local
  ERROR invalid lxd-profile.yaml: contains device type "unix-disk"
$ juju deploy  ./testcharms/charm-repo/quantal/lxd-profile-fail lxd-profilefail-local --force
  Deploying charm "local:bionic/lxd-profile-fail-0"
$ juju deploy ghost --series bionic
  ERROR series "bionic" not supported by charm, supported series are: xenial,trusty. Use --force to deploy the charm anyway.
$ juju upgrade-charm lxd-profile-local --path ./testcharms/charm-repo/quantal/lxd-profile
  ERROR invalid lxd-profile.yaml: contains config value "limits.memory"
$ juju upgrade-charm lxd-profile-local --path ./testcharms/charm-repo/quantal/lxd-profile --force
  Added charm "local:bionic/lxd-profile-2" to the model.

Do we want to add "Use --force to deploy the charm anyway." to the Error message when a profile validation fails?

File: cmd/juju/application/upgradecharm.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/juju/application/upgradecharm.go]: I agree with @mitechie on this one. I think we can clean this up in the future. I'll leave a comment.

**Comment by SimonRichardson** [comment]
> Do we want to add "Use --force to deploy the charm anyway." to the Error message when a profile validation fails?

I'd rather not, that implies that people will just think it's ok to use `--force`. When actually we want `--force` to be used in limited scenarios, otherwise we might as well just run `--force` all the time...

File: api/client_test.go [hunk 5]
```diff
@@ -172,56 +180,96 @@ func (s *clientSuite) TestAddLocalCharmNoHooks(c *gc.C) {
...
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(savedURL.Revision, gc.Equals, 42)
+
+	// Upload a charm directory again, revision should be bumped.
+	savedURL, err = client.AddLocalCharm(curl, charmDir, true)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(savedURL.String(), gc.Equals, curl.WithRevision(43).String())
+}
 
+func (s *clientSuite) TestAddLocalCharmWithInvalidLXDProfileWithForceSucceeds(c *gc.C) {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply api/client_test.go]: Done with latest commit #5be3a6f

File: juju/testing/conn.go [hunk 6]
```diff
@@ -676,6 +678,10 @@ func AddCharm(st *state.State, curl *charm.URL, ch charm.Charm) (*state.Charm, e
 		return nil, err
 	}
 
+	if err := lxdprofile.ValidateCharmLXDProfile(ch); err != nil && !force {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply juju/testing/conn.go]: Resolved in latest commit #5be3a6f

File: apiserver/facades/client/application/charmstore.go [hunk 7]
```diff
@@ -140,7 +140,11 @@ func AddCharmWithAuthorizationAndRepo(st State, args params.AddCharmWithAuthoriz
 
 	// Validate the charm lxd profile once we've downloaded it.
 	if err := lxdprofile.ValidateCharmLXDProfile(downloadedCharm); err != nil {
-		return errors.Annotate(err, "cannot add charm")
+		if args.Force {
+			logger.Debugf("force flag used to override validation error %v", err)
```
Comment (hmlanigan) [inline apiserver/facades/client/application/charmstore.go]: This debug msg is duplicated somewhere else and get's printed twice.

Comment (SimonRichardson) [reply apiserver/facades/client/application/charmstore.go]: I'll see what I can do, as we don't pass the error back stating this is a validation error, so we have to debug at a lower level.

File: cmd/juju/application/bundle_test.go [hunk 8]
```diff
@@ -865,6 +873,55 @@ func (s *BundleDeployCharmStoreSuite) TestDeployBundleLocalDeploymentBadConfig(c
...
+        series: bionic
+        services:
+            lxd-profile-fail:
+                charm: %s
+                num_units: 1
+    `, lxdProfilePath))
+	c.Assert(err, gc.ErrorMatches, "cannot deploy bundle: cannot deploy local charm at .*: invalid lxd-profile.yaml: contains device type \"unix-disk\"")
+}
+
+func (s *BundleDeployCharmStoreSuite) TestDeployBundleLocalDeploymentBadLXDProfileWithForce(c *gc.C) {
```
Comment (hmlanigan) [inline cmd/juju/application/bundle_test.go]: Do we still need this test?

File: cmd/juju/application/deploy.go [hunk 9]
```diff
@@ -623,7 +633,7 @@ var (
 // whether we are deploying a charm or a bundle.
 func charmOnlyFlags() []string {
 	charmOnlyFlags := []string{
-		"bind", "config", "constraints", "force", "n", "num-units",
+		"bind", "config", "constraints", "n", "num-units",
```
Comment (hmlanigan) [inline cmd/juju/application/deploy.go]: don't we need "force" back in this slice?

File: cmd/juju/application/validate.go (hunk 1 above)
Comment (hmlanigan) [inline cmd/juju/application/validate.go]: Remove TODO?  it's been addressed.

File: testcharms/charm-repo/quantal/lxd-profile-fail/lxd-profile.yaml [hunk 10]
```diff
@@ -5,6 +5,7 @@ config:
   linux.kernel_modules: openvswitch,nbd,ip_tables,ip6_tables
   # boot.* keys are blacklisted by juju.
   boot.autostart: "true"
+  limits.memory: 256mb
```
Comment (hmlanigan) [inline testcharms/charm-repo/quantal/lxd-profile-fail/lxd-profile.yaml]: Causes profile to fail even with force:
cannot write lxd profiles: Invalid value: 256mb

File: cmd/juju/application/bundle_test.go (hunk 8 above)
Comment (SimonRichardson) [reply cmd/juju/application/bundle_test.go]: I want it to ensure people don't do the same mistake as me in the future :D

File: cmd/juju/application/deploy.go (hunk 9 above)
Comment (SimonRichardson) [reply cmd/juju/application/deploy.go]: Yes!

File: cmd/juju/application/validate.go (hunk 1 above)
Comment (SimonRichardson) [reply cmd/juju/application/validate.go]: Good spot.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**